### PR TITLE
changes in fsdp evaluation 

### PR DESCRIPTION
**Pull request by humaira-rf**
## Changes
- Downsampled fsdp large and fsdp medium notebooks for demo purposes.

- Increased eval batch size for fsdp-large notebook
- Changed fsdp evaluation such that each GPU's model.generate() will handle different batch.

## Tested
Tested on SFT - FSDP and non-FSDP Lite Notebooks

### CODE REVIEW

File: rapidfireai/fit/ml/callbacks.py [hunk 1]
```diff
@@ -262,34 +316,104 @@ def _compute_metrics_fsdp(
...
+        """
         rank = dist.get_rank() if dist.is_initialized() else 0
+        world_size = dist.get_world_size() if dist.is_initialized() else 1
+
+        num_samples = input_ids.shape[0]
 
-        # All ranks MUST call model.generate() together -- FSDP forward
-        # uses all-gather/reshard which are collective NCCL operations.
+        # ---- shard the eval data across ranks ----
+        shard_size = (num_samples + world_size - 1) // world_size  # ceil div
```
Comment (arun-rfai) [inline rapidfireai/fit/ml/callbacks.py]: This formula seems strange. Isn't shard_size just per_device_eval_batch_size? Why are you doing - 1 and ceil and such? Even so, this seems buggy. Let's say there are 64 samples and world_size = 4. Then shard_size = 16. But this formula gives 17, which is incorrect.

If the eval dataset size is NOT a multiple of per_device_eval_batch_size * num_gpus, typically it is padded with 0s as dummy examples on the last minibatch to add up to a multiple. Is this not done by TRL?

File: tutorial_notebooks/fine-tuning/rf-tutorial-sft-chatqa-fsdp.ipynb [hunk 2]
```diff
@@ -41,8 +41,8 @@
         "\n",
         "dataset=load_dataset(\"bitext/Bitext-customer-support-llm-chatbot-training-dataset\")\n",
         "\n",
-        "train_dataset=dataset[\"train\"].select(range(2000))\n",
-        "eval_dataset=dataset[\"train\"].select(range(2000,2050))\n",
+        "train_dataset=dataset[\"train\"].select(range(512))\n",
```
Comment (arun-rfai) [inline tutorial_notebooks/fine-tuning/rf-tutorial-sft-chatqa-fsdp.ipynb]: Why not make this 320 examples as well like the FSDP LARGE notebook? Helps reduce runtime.

**Review by arun-rfai — CHANGES_REQUESTED** [review]

File: rapidfireai/fit/ml/callbacks.py (hunk 1 above)
Comment (humaira-rf) [reply rapidfireai/fit/ml/callbacks.py]: shard size means how much total data(upper limit) each gpu is gonna deal within single GPU for a single eval step: so if there are 64 samples and 4 gpus, each GPU will get (64+4-1)//4=16(not 17) samples total and if per_device_eval_batch_size is 4, there will be 4 passes on samples of size 4.

If there are 65 eval samples total, each GPU should deal with atmost 65+4-1/4=17
so GPU1 - 17, GPU2 - 17, GPU 3 - 17, GPU 4 - 14 and we will add 3 extra dummy samples to GPU4
In order for the  communication collectives to work, all gpu's must have same size of data. So This ceil tries to give easy logic of equalizing sample sizes.  If we instead have 16,16,16,17 - that would mean first three GPUs' datasets need to be padded with single sample to equalize to 17(similar logic - different implementation).  Our way is kind of similar to DDP Logic.

Comment (arun-rfai) [reply rapidfireai/fit/ml/callbacks.py]: Oh I see, makes sense. The comment on the code threw me off--it is *floor div*, not "ceil div". I like your approach of padding at most one microbatch. I see the padding code also removes the copied samples later. So, everything checks out now.

File: rapidfireai/fit/ml/callbacks.py [hunk 3]
```diff
@@ -321,20 +321,25 @@ def _compute_metrics_fsdp(
...
         world_size = dist.get_world_size() if dist.is_initialized() else 1
 
-        num_samples = input_ids.shape[0]
+        total_eval_samples = input_ids.shape[0]
 
         # ---- shard the eval data across ranks ----
-        shard_size = (num_samples + world_size - 1) // world_size  # ceil div
-        start_idx = rank * shard_size
-        end_idx = min(start_idx + shard_size, num_samples)
+        samples_per_gpu = (total_eval_samples + world_size - 1) // world_size  # ceil div
```
Comment (arun-rfai) [inline rapidfireai/fit/ml/callbacks.py:327]: // is "floor div", not "ceil div"

**Review by arun-rfai — DISMISSED** [review]
LGTM! 
(Minor nit: "ceil div" -> "floor div" in callbacks.py)
